transformers/recommendations/matrixfactorization.py

```diff
@@ -0,0 +1,136 @@
+"""Collaborative filtering features using various techniques of Matrix Factorization for recommendations"""
+import datatable as dt
+import numpy as np
+import pandas as pd
+
+import scipy
+import h2o4gpu
+
+from h2oaicore.transformer_utils import CustomTransformer
+from sklearn.decomposition import NMF
+from sklearn.model_selection import KFold
+from sklearn.preprocessing import LabelEncoder
+
+
+class RecH2OMFTransformer(CustomTransformer):
+    _multiclass = False
+    _can_use_gpu = True
+    _mf_type = "h2o4gpu"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.user_col = "user" ## edit this to the user column name
+        self.item_col = "item" ## edit this to the item column name
+
+    @staticmethod
+    def do_acceptance_test():
+        return False
+
+    @staticmethod
+    def get_default_properties():
+        return dict(col_type="all", min_cols="all", max_cols="all", relative_importance=1, num_default_instances=1)
+
+    def set_params(self):
+        if self.__class__._mf_type == "h2o4gpu":
+            self.params = {"n_components": 50,
+                           "lambda": 0.01,
+                           "batches": 1,
+                           "max_iter": 50}
+        elif self.__class__._mf_type == "nmf":
+            self.params = {"n_components": 50,
+                           "alpha": 0.01,
+                           "max_iter": 50}
+
+    def fit_transform(self, X: dt.Frame, y: np.array = None):
+        self.set_params()
+
+        if len(np.unique(self.labels)) == 2:
+            le = LabelEncoder()
+            self.labels = le.fit_transform(self.labels)
+            y = np.array(le.transform(y), dtype="float32")
+        else:
+            y = np.array(y, dtype="float32")
+
+        X = X[:, [self.user_col, self.item_col]]
+
+        self.user_le = LabelEncoder()
+        self.item_le = LabelEncoder()
+
+        X[:, self.user_col] = dt.Frame(self.user_le.fit_transform(X[:, self.user_col]))
```

Comment:
probably safer to pass a numpy array to the LabelEncoder(), rather than a dt.Frame